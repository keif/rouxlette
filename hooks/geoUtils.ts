/**
 * Geographic utility functions for location calculations
 */

export type Coords = { latitude: number; longitude: number };

/**
 * Calculate the haversine distance between two coordinate points in kilometers
 * Uses the haversine formula for great-circle distance calculation
 * 
 * @param a - First coordinate point
 * @param b - Second coordinate point
 * @returns Distance in kilometers
 */
export function haversineKm(a: Coords, b: Coords): number {
	const R = 6371; // Earth's radius in kilometers
	
	const dLat = ((b.latitude - a.latitude) * Math.PI) / 180;
	const dLon = ((b.longitude - a.longitude) * Math.PI) / 180;
	
	const lat1 = (a.latitude * Math.PI) / 180;
	const lat2 = (b.latitude * Math.PI) / 180;
	
	const x = Math.sin(dLat / 2) ** 2 + 
		Math.sin(dLon / 2) ** 2 * Math.cos(lat1) * Math.cos(lat2);
	
	return 2 * R * Math.asin(Math.min(1, Math.sqrt(x)));
}

/**
 * Find the closest result to a reference point by haversine distance
 * 
 * @param results - Array of geocoding results with geometry.location
 * @param referencePoint - Coordinate to compare distances to
 * @returns The closest result or null if no results
 */
export function findClosestResult(
	results: any[], 
	referencePoint: Coords
): any | null {
	if (!results?.length) return null;
	
	let closest: any = null;
	let minDistance = Infinity;
	
	for (const result of results) {
		const location = result?.geometry?.location;
		if (!location?.lat || !location?.lng) continue;
		
		const coords = { latitude: location.lat, longitude: location.lng };
		const distance = haversineKm(referencePoint, coords);
		
		if (distance < minDistance) {
			minDistance = distance;
			closest = result;
		}
	}
	
	return closest;
}

/**
 * Extract canonical location label from geocoding result
 * Returns format like "Powell, OH" or "Columbus, OH"
 * 
 * @param result - Geocoding result with address_components
 * @returns Canonical location string
 */
export function extractCanonicalLabel(result: any): string {
	try {
		const components = result?.address_components || [];
		
		// Find locality (city/town)
		const locality = components.find((c: any) => 
			c.types?.includes('locality')
		);
		
		// Find state (administrative_area_level_1)  
		const state = components.find((c: any) => 
			c.types?.includes('administrative_area_level_1')
		);
		
		if (locality && state) {
			return `${locality.long_name}, ${state.short_name}`;
		}
		
		// Fallback to administrative_area_level_2 (county) if no locality
		const county = components.find((c: any) => 
			c.types?.includes('administrative_area_level_2')
		);
		
		if (county && state) {
			return `${county.long_name}, ${state.short_name}`;
		}
		
		// Last resort - use formatted address
		return result.formatted_address || 'Unknown Location';
		
	} catch (error) {
		return 'Unknown Location';
	}
}

/**
 * Extract state code from geocoding result
 * 
 * @param result - Geocoding result with address_components
 * @returns State code (e.g., "OH") or null
 */
export function extractStateFromResult(result: any): string | null {
	try {
		const components = result?.address_components || [];
		const state = components.find((c: any) => 
			c.types?.includes('administrative_area_level_1')
		);
		return state?.short_name || null;
	} catch (error) {
		return null;
	}
}

/**
 * Extract coordinates from geocoding result
 * 
 * @param result - Geocoding result with geometry.location
 * @returns Coordinates or null
 */
export function extractCoordsFromResult(result: any): Coords | null {
	try {
		const location = result?.geometry?.location;
		if (location?.lat && location?.lng) {
			return { latitude: location.lat, longitude: location.lng };
		}
		return null;
	} catch (error) {
		return null;
	}
}