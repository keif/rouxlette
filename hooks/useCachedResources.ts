import { FontAwesome } from '@expo/vector-icons';
import * as Font from 'expo-font';
import * as SplashScreen from 'expo-splash-screen';
import { useEffect, useState } from 'react';
import { logSafe } from '../utils/log';

export default function useCachedResources() {
  const [isLoadingComplete, setLoadingComplete] = useState(false);

  // Load any resources or data that we need prior to rendering the app
  useEffect(() => {
    async function loadResourcesAndDataAsync() {
      try {
        SplashScreen.preventAutoHideAsync();

        // Load fonts
        await Font.loadAsync({
          ...FontAwesome.font,
          'space-mono': require('../assets/fonts/SpaceMono-Regular.ttf'),
          'Roboto-Bold': require('../assets/fonts/Roboto-Bold.ttf'),
          'Roboto-Medium': require('../assets/fonts/Roboto-Medium.ttf'),
          'Roboto-Regular': require('../assets/fonts/Roboto-Regular.ttf'),
          'WorkSans-Bold': require('../assets/fonts/WorkSans-Bold.ttf'),
          'WorkSans-Medium': require('../assets/fonts/WorkSans-Medium.ttf'),
          'WorkSans-Regular': require('../assets/fonts/WorkSans-Regular.ttf'),
          'WorkSans-SemiBold': require('../assets/fonts/WorkSans-SemiBold.ttf'),
        });
      } catch (e: any) {
        // We might want to provide this error information to an error reporting service
        logSafe('Font loading error', { message: e?.message });
      } finally {
        setLoadingComplete(true);
        SplashScreen.hideAsync();
      }
    }

    loadResourcesAndDataAsync();
  }, []);

  return isLoadingComplete;
}
