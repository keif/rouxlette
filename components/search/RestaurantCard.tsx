import React, { useEffect, useRef } from "react";
import { Animated, Image, StyleSheet, Text, useWindowDimensions, View, Linking, Pressable, Platform } from "react-native";
import { BusinessProps } from "../../hooks/useResults";
import AppStyles from "../../AppStyles";
import Icon from "react-native-vector-icons/MaterialIcons";
import { FontAwesome, MaterialIcons } from "@expo/vector-icons";
import StarRating from "../shared/StarRating";
import OpenSign from "../results/OpenSign";
import FlipCard from "../shared/FlipCard";
import Config from "../../Config";

interface RestaurantCardProps {
	index: number;
	result: BusinessProps;
}

const RestaurantCard = ({ index, result }: RestaurantCardProps) => {
	const { categories, hours, image_url, is_closed, location, name, price, rating, review_count, url, phone, display_phone, photos } = result;
	const { width } = useWindowDimensions();
	const translateY = useRef<Animated.Value>(new Animated.Value(50)).current;
	const opacity = useRef<Animated.Value>(new Animated.Value(0)).current;
	const is_open_now = hours && hours[0]?.is_open_now || false;

	useEffect(() => {
		Animated.parallel([
			Animated.timing(translateY, {
				toValue: 0,
				duration: 400,
				delay: (400 / 3),
				useNativeDriver: true,
			}),
			Animated.timing(opacity, {
				toValue: 1,
				duration: 400,
				delay: (400 / 3),
				useNativeDriver: true,
			}),
		]).start();
	}, []);

	const handleYelpPress = () => {
		Linking
			.openURL(url)
			.catch((err) => console.error("An error occurred:", err));
	};

	const handlePhonePress = () => {
		if (phone) {
			Linking.openURL(`tel:${phone}`);
		}
	};

	const handleMapsPress = () => {
		const address = location.display_address.join(', ');
		const encodedAddress = encodeURIComponent(address);
		
		if (Platform.OS === 'ios') {
			Linking.openURL(`http://maps.apple.com/?q=${encodedAddress}`);
		} else {
			Linking.openURL(`https://www.google.com/maps/search/?api=1&query=${encodedAddress}`);
		}
	};

	const imageSize = width - 24;

	const frontContent = (
		<View style={styles.cardContent}>
			<View style={styles.imageContainer}>
				<Image
					style={{ height: imageSize / 2, width: imageSize }}
					source={{ uri: image_url }}
					resizeMode="cover"
				/>
				<Text style={styles.index}>{index + 1}.</Text>
				<Icon
					style={styles.iconFavorite}
					name="favorite-border"
					size={24}
					color={AppStyles.color.primary}
				/>
			</View>
			<View style={styles.detail}>
				<View style={styles.detailHeader}>
					<Text style={styles.name}>{name}</Text>
					<Text style={{ fontSize: 22, fontFamily: AppStyles.fonts.semiBold, }}>
						{price}
					</Text>
				</View>
				<View style={{ flexDirection: "row" }}>
					<Text style={styles.subText} numberOfLines={1}>
						{categories.map(cat => cat.title).join(`, `)} â€¢ {location.city}
					</Text>
				</View>
				<View style={{ flexDirection: "row", marginTop: 4 }}>
					<StarRating rating={rating} />
					<Text style={styles.review}>{review_count} Reviews</Text>
				</View>
			</View>
		</View>
	);

	const backContent = (
		<View style={styles.cardContent}>
			<View style={styles.backHeader}>
				<Text style={styles.backTitle}>{name}</Text>
				<OpenSign is_open_now={is_open_now} />
			</View>
			
			<View style={styles.backRating}>
				<StarRating rating={rating} />
				<Text style={styles.backReviewText}>{review_count} Review{review_count > 1 ? 's' : ''}</Text>
			</View>

			<View style={styles.backDetails}>
				<Text style={styles.backDetailText}>
					<MaterialIcons name="location-on" size={16} color={AppStyles.color.primary} />
					{' '}{location.display_address.join(', ')}
				</Text>
				
				{phone && (
					<Text style={styles.backDetailText}>
						<MaterialIcons name="phone" size={16} color={AppStyles.color.primary} />
						{' '}{display_phone}
					</Text>
				)}

				<Text style={styles.backDetailText}>
					<MaterialIcons name="category" size={16} color={AppStyles.color.primary} />
					{' '}{categories.map(cat => cat.title).join(', ')}
				</Text>
			</View>

			<View style={styles.backActions}>
				<Pressable
					style={({ pressed }) => [
						styles.backButton,
						{ opacity: !Config.isAndroid && pressed ? 0.6 : 1 },
					]}
					onPress={handleMapsPress}
					android_ripple={{
						color: "grey",
						radius: 28,
						borderless: true,
					}}
				>
					<MaterialIcons
						color={AppStyles.color.primary}
						name="map"
						size={20}
					/>
					<Text style={styles.backButtonText}>Maps</Text>
				</Pressable>

				<Pressable
					style={({ pressed }) => [
						styles.backButton,
						{ opacity: !Config.isAndroid && pressed ? 0.6 : 1 },
					]}
					onPress={handleYelpPress}
					android_ripple={{
						color: "grey",
						radius: 28,
						borderless: true,
					}}
				>
					<FontAwesome
						color={AppStyles.color.yelp}
						name="yelp"
						size={20}
					/>
					<Text style={styles.backButtonText}>Yelp</Text>
				</Pressable>

				{phone && (
					<Pressable
						style={({ pressed }) => [
							styles.backButton,
							{ opacity: !Config.isAndroid && pressed ? 0.6 : 1 },
						]}
						onPress={handlePhonePress}
						android_ripple={{
							color: "grey",
							radius: 28,
							borderless: true,
						}}
					>
						<MaterialIcons
							color={AppStyles.color.phone}
							name="phone-in-talk"
							size={20}
						/>
						<Text style={styles.backButtonText}>Call</Text>
					</Pressable>
				)}
			</View>
		</View>
	);

	return (
		<Animated.View
			style={[styles.container, { opacity, transform: [{ translateY }] }]}
		>
			<FlipCard
				front={frontContent}
				back={backContent}
				style={styles.flipCard}
			/>
		</Animated.View>
	);
};

const textStyle = {
	color: "rgba(128,128,128, 0.80)",
	fontFamily: AppStyles.fonts.regular,
};

const styles = StyleSheet.create({
	container: {
		marginHorizontal: 12,
		marginVertical: 12,
	},
	flipCard: {
		backgroundColor: AppStyles.color.white,
		borderRadius: 16,
		elevation: 8,
		shadowColor: AppStyles.color.shadow,
		shadowOffset: {
			height: 4,
			width: 4,
		},
		shadowOpacity: 0.3,
		shadowRadius: 12,
	},
	cardContent: {
		borderRadius: 16,
		overflow: 'hidden',
	},
	detail: {
		paddingHorizontal: 16,
		paddingVertical: 8,
	},
	detailHeader: {
		flex: 1,
		flexDirection: "row",
	},
	iconFavorite: {
		padding: 16,
		position: "absolute",
		right: 0,
	},
	imageContainer: {
		borderTopLeftRadius: 16,
		borderTopRightRadius: 16,
		overflow: "hidden",
	},
	index: {
		color: AppStyles.color.white,
		fontSize: 22,
		fontFamily: AppStyles.fonts.semiBold,
		fontWeight: "bold",
		padding: 16,
		position: "absolute",
		left: 0,
		textShadowColor: AppStyles.color.black,
		textShadowOffset: {
			width: 0,
			height: 0,
		},
		textShadowRadius: 6,
	},
	name: {
		color: AppStyles.color.black,
		flex: 1,
		fontSize: 22,
		fontFamily: AppStyles.fonts.semiBold,
		fontWeight: "bold",
		paddingBottom: 6,
	},
	subText: {
		...textStyle,
		flex: 1,
		paddingRight: 4,
	},
	review: {
		...textStyle,
		marginLeft: 8,
	},
	// Back of card styles
	backHeader: {
		flexDirection: 'row',
		justifyContent: 'space-between',
		alignItems: 'center',
		padding: 16,
		backgroundColor: AppStyles.color.primary,
	},
	backTitle: {
		color: AppStyles.color.white,
		fontSize: 20,
		fontFamily: AppStyles.fonts.semiBold,
		fontWeight: 'bold',
		flex: 1,
	},
	backRating: {
		flexDirection: 'row',
		alignItems: 'center',
		paddingHorizontal: 16,
		paddingTop: 12,
	},
	backReviewText: {
		...textStyle,
		marginLeft: 8,
	},
	backDetails: {
		paddingHorizontal: 16,
		paddingVertical: 12,
	},
	backDetailText: {
		...textStyle,
		fontSize: 16,
		marginBottom: 8,
		lineHeight: 22,
	},
	backActions: {
		flexDirection: 'row',
		justifyContent: 'space-around',
		paddingVertical: 16,
		paddingHorizontal: 16,
	},
	backButton: {
		backgroundColor: 'rgba(255, 255, 255, 0.9)',
		flexDirection: 'row',
		alignItems: 'center',
		paddingHorizontal: 12,
		paddingVertical: 8,
		borderRadius: 20,
		shadowColor: AppStyles.input.shadow,
		...AppStyles.ButtonPressable,
		shadowOpacity: 0.2,
		shadowRadius: 3,
		elevation: 3,
	},
	backButtonText: {
		marginLeft: 6,
		fontSize: 14,
		fontFamily: AppStyles.fonts.medium,
		color: AppStyles.color.black,
	},
});

export default RestaurantCard;