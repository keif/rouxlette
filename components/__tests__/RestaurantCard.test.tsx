import * as React from 'react';
import renderer from 'react-test-renderer';

import RestaurantCard from '../search/RestaurantCard';
import { BusinessProps } from '../../hooks/useResults';

// Mock react-native-reanimated
jest.mock('react-native-reanimated', () => {
  const Reanimated = require('react-native-reanimated/mock');
  Reanimated.default.call = () => {};
  return Reanimated;
});

// Mock Animated from react-native
jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');
  return {
    ...RN,
    Animated: {
      ...RN.Animated,
      parallel: jest.fn(() => ({ start: jest.fn() })),
      timing: jest.fn(() => ({ start: jest.fn() })),
      Value: jest.fn(() => ({
        setValue: jest.fn(),
        addListener: jest.fn(),
        removeListener: jest.fn(),
        hasListeners: jest.fn(),
        stopAnimation: jest.fn(),
        resetAnimation: jest.fn(),
        interpolate: jest.fn(),
        extractOffset: jest.fn(),
        setOffset: jest.fn(),
        flattenOffset: jest.fn(),
        removeClampedListener: jest.fn(),
        addClampedListener: jest.fn(),
      })),
    },
  };
});

// Mock react-native-gesture-handler
jest.mock('react-native-gesture-handler', () => {
  const original = jest.requireActual('react-native-gesture-handler');
  return {
    ...original,
    TapGestureHandler: ({ children }: any) => children,
    PanGestureHandler: ({ children }: any) => children,
    State: {
      END: 4,
    },
  };
});

// Mock vector icons
jest.mock('react-native-vector-icons/MaterialIcons', () => 'Icon');
jest.mock('@expo/vector-icons', () => ({
  FontAwesome: 'FontAwesome',
  MaterialIcons: 'MaterialIcons',
}));

// Mock Config
jest.mock('../../Config', () => ({
  isAndroid: false,
}));

const mockBusiness: BusinessProps = {
  id: 'test-id',
  name: 'Test Restaurant',
  image_url: 'https://example.com/image.jpg',
  url: 'https://yelp.com/test-restaurant',
  phone: '+15551234567',
  display_phone: '(555) 123-4567',
  review_count: 150,
  categories: [
    { alias: 'italian', title: 'Italian' },
    { alias: 'pizza', title: 'Pizza' }
  ],
  rating: 4.5,
  coordinates: {
    latitude: 37.7749,
    longitude: -122.4194
  },
  transactions: [],
  price: '$$',
  location: {
    address1: '123 Main St',
    address2: '',
    address3: '',
    city: 'San Francisco',
    zip_code: '94102',
    country: 'US',
    state: 'CA',
    display_address: ['123 Main St', 'San Francisco, CA 94102']
  },
  hours: [{
    open: [],
    hours_type: 'REGULAR',
    is_open_now: true
  }],
  is_closed: false,
  photos: ['https://example.com/photo1.jpg', 'https://example.com/photo2.jpg']
};

it('renders RestaurantCard correctly', () => {
  const tree = renderer.create(
    React.createElement(RestaurantCard, { index: 0, result: mockBusiness })
  ).toJSON();

  expect(tree).toMatchSnapshot();
});