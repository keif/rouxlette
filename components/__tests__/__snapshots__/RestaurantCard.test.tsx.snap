// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders RestaurantCard correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "marginHorizontal": 12,
      "marginVertical": 12,
      "opacity": 0,
      "transform": [
        {
          "translateY": 50,
        },
      ],
    }
  }
>
  <View
    accessibilityHint="Flip card to see more details"
    accessibilityRole="button"
    accessible={true}
    style={
      [
        {
          "position": "relative",
        },
        {
          "backgroundColor": "#fff",
          "borderRadius": 16,
          "elevation": 8,
          "shadowColor": "#333",
          "shadowOffset": {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 12,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "height": "100%",
            "position": "absolute",
            "width": "100%",
          },
          {
            "backfaceVisibility": "hidden",
            "opacity": undefined,
            "pointerEvents": "auto",
            "transform": [
              {
                "perspective": 1000,
              },
              {
                "rotateY": "undefineddeg",
              },
            ],
          },
        ]
      }
    >
      <View
        style={
          {
            "borderRadius": 16,
            "overflow": "hidden",
          }
        }
      >
        <View
          style={
            {
              "borderTopLeftRadius": 16,
              "borderTopRightRadius": 16,
              "overflow": "hidden",
            }
          }
        >
          <Image
            resizeMode="cover"
            source={
              {
                "uri": "https://example.com/image.jpg",
              }
            }
            style={
              {
                "height": 363,
                "width": 726,
              }
            }
          />
          <Text
            style={
              {
                "color": "#fff",
                "fontFamily": "WorkSans-SemiBold",
                "fontSize": 22,
                "fontWeight": "bold",
                "left": 0,
                "padding": 16,
                "position": "absolute",
                "textShadowColor": "#000",
                "textShadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "textShadowRadius": 6,
              }
            }
          >
            1
            .
          </Text>
          <Icon
            color="rgba(51,178,73,1)"
            name="favorite-border"
            size={24}
            style={
              {
                "padding": 16,
                "position": "absolute",
                "right": 0,
              }
            }
          />
        </View>
        <View
          style={
            {
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            }
          }
        >
          <View
            style={
              {
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                {
                  "color": "#000",
                  "flex": 1,
                  "fontFamily": "WorkSans-SemiBold",
                  "fontSize": 22,
                  "fontWeight": "bold",
                  "paddingBottom": 6,
                }
              }
            >
              Test Restaurant
            </Text>
            <Text
              style={
                {
                  "fontFamily": "WorkSans-SemiBold",
                  "fontSize": 22,
                }
              }
            >
              $$
            </Text>
          </View>
          <View
            style={
              {
                "flexDirection": "row",
              }
            }
          >
            <Text
              numberOfLines={1}
              style={
                {
                  "color": "rgba(128,128,128, 0.80)",
                  "flex": 1,
                  "fontFamily": "WorkSans-Regular",
                  "paddingRight": 4,
                }
              }
            >
              Italian, Pizza
              <View
                style={
                  {
                    "width": 4,
                  }
                }
              />
              <Icon
                color="rgba(51,178,73,1)"
                name="location-pin"
                size={12}
              />
              San Francisco
            </Text>
          </View>
          <View
            style={
              {
                "flexDirection": "row",
                "marginTop": 4,
              }
            }
          >
            <View
              style={
                {
                  "display": "flex",
                  "flexDirection": "row",
                }
              }
            >
              <MaterialIcons
                name="star"
                size={20}
                style={
                  {
                    "color": "#ffb300",
                  }
                }
              />
              <MaterialIcons
                name="star"
                size={20}
                style={
                  {
                    "color": "#ffb300",
                  }
                }
              />
              <MaterialIcons
                name="star"
                size={20}
                style={
                  {
                    "color": "#ffb300",
                  }
                }
              />
              <MaterialIcons
                name="star"
                size={20}
                style={
                  {
                    "color": "#ffb300",
                  }
                }
              />
              <MaterialIcons
                name="star-half"
                size={20}
                style={
                  {
                    "color": "#ffb300",
                  }
                }
              />
            </View>
            <Text
              style={
                {
                  "color": "rgba(128,128,128, 0.80)",
                  "fontFamily": "WorkSans-Regular",
                  "marginLeft": 8,
                }
              }
            >
              150
               Reviews
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "height": "100%",
            "position": "absolute",
            "width": "100%",
          },
          {
            "backfaceVisibility": "hidden",
            "opacity": undefined,
            "pointerEvents": "none",
            "transform": [
              {
                "perspective": 1000,
              },
              {
                "rotateY": "undefineddeg",
              },
            ],
          },
        ]
      }
    >
      <View
        style={
          {
            "borderRadius": 16,
            "overflow": "hidden",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(51,178,73,1)",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "padding": 16,
            }
          }
        >
          <Text
            style={
              {
                "color": "#fff",
                "flex": 1,
                "fontFamily": "WorkSans-SemiBold",
                "fontSize": 20,
                "fontWeight": "bold",
              }
            }
          >
            Test Restaurant
          </Text>
          <Text
            style={
              {
                "color": "#71C562",
                "fontSize": 18,
                "fontWeight": "bold",
                "textShadowColor": "#000",
                "textShadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "textShadowRadius": 6,
              }
            }
          >
            Open
          </Text>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "paddingHorizontal": 16,
              "paddingTop": 12,
            }
          }
        >
          <View
            style={
              {
                "display": "flex",
                "flexDirection": "row",
              }
            }
          >
            <MaterialIcons
              name="star"
              size={20}
              style={
                {
                  "color": "#ffb300",
                }
              }
            />
            <MaterialIcons
              name="star"
              size={20}
              style={
                {
                  "color": "#ffb300",
                }
              }
            />
            <MaterialIcons
              name="star"
              size={20}
              style={
                {
                  "color": "#ffb300",
                }
              }
            />
            <MaterialIcons
              name="star"
              size={20}
              style={
                {
                  "color": "#ffb300",
                }
              }
            />
            <MaterialIcons
              name="star-half"
              size={20}
              style={
                {
                  "color": "#ffb300",
                }
              }
            />
          </View>
          <Text
            style={
              {
                "color": "rgba(128,128,128, 0.80)",
                "fontFamily": "WorkSans-Regular",
                "marginLeft": 8,
              }
            }
          >
            150
             Review
            s
          </Text>
        </View>
        <View
          style={
            {
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
        >
          <Text
            style={
              {
                "color": "rgba(128,128,128, 0.80)",
                "fontFamily": "WorkSans-Regular",
                "fontSize": 16,
                "lineHeight": 22,
                "marginBottom": 8,
              }
            }
          >
            <MaterialIcons
              color="rgba(51,178,73,1)"
              name="location-on"
              size={16}
            />
             
            123 Main St, San Francisco, CA 94102
          </Text>
          <Text
            style={
              {
                "color": "rgba(128,128,128, 0.80)",
                "fontFamily": "WorkSans-Regular",
                "fontSize": 16,
                "lineHeight": 22,
                "marginBottom": 8,
              }
            }
          >
            <MaterialIcons
              color="rgba(51,178,73,1)"
              name="phone"
              size={16}
            />
             
            (555) 123-4567
          </Text>
          <Text
            style={
              {
                "color": "rgba(128,128,128, 0.80)",
                "fontFamily": "WorkSans-Regular",
                "fontSize": 16,
                "lineHeight": 22,
                "marginBottom": 8,
              }
            }
          >
            <MaterialIcons
              color="rgba(51,178,73,1)"
              name="category"
              size={16}
            />
             
            Italian, Pizza
          </Text>
        </View>
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-around",
              "paddingHorizontal": 16,
              "paddingVertical": 16,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "rgba(255, 255, 255, 0.9)",
                  "borderRadius": 36,
                  "elevation": 3,
                  "flexDirection": "row",
                  "fontSize": 30,
                  "padding": 16,
                  "paddingHorizontal": 12,
                  "paddingVertical": 8,
                  "shadowColor": "#333",
                  "shadowOffset": {
                    "height": 4,
                    "width": 0,
                  },
                  "shadowOpacity": 0.2,
                  "shadowRadius": 3,
                },
                {
                  "opacity": 1,
                },
              ]
            }
          >
            <MaterialIcons
              color="rgba(51,178,73,1)"
              name="map"
              size={20}
            />
            <Text
              style={
                {
                  "color": "#000",
                  "fontFamily": "WorkSans-Medium",
                  "fontSize": 14,
                  "marginLeft": 6,
                }
              }
            >
              Maps
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "rgba(255, 255, 255, 0.9)",
                  "borderRadius": 36,
                  "elevation": 3,
                  "flexDirection": "row",
                  "fontSize": 30,
                  "padding": 16,
                  "paddingHorizontal": 12,
                  "paddingVertical": 8,
                  "shadowColor": "#333",
                  "shadowOffset": {
                    "height": 4,
                    "width": 0,
                  },
                  "shadowOpacity": 0.2,
                  "shadowRadius": 3,
                },
                {
                  "opacity": 1,
                },
              ]
            }
          >
            <FontAwesome
              color="#c41200"
              name="yelp"
              size={20}
            />
            <Text
              style={
                {
                  "color": "#000",
                  "fontFamily": "WorkSans-Medium",
                  "fontSize": 14,
                  "marginLeft": 6,
                }
              }
            >
              Yelp
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "rgba(255, 255, 255, 0.9)",
                  "borderRadius": 36,
                  "elevation": 3,
                  "flexDirection": "row",
                  "fontSize": 30,
                  "padding": 16,
                  "paddingHorizontal": 12,
                  "paddingVertical": 8,
                  "shadowColor": "#333",
                  "shadowOffset": {
                    "height": 4,
                    "width": 0,
                  },
                  "shadowOpacity": 0.2,
                  "shadowRadius": 3,
                },
                {
                  "opacity": 1,
                },
              ]
            }
          >
            <MaterialIcons
              color="#71C562"
              name="phone-in-talk"
              size={20}
            />
            <Text
              style={
                {
                  "color": "#000",
                  "fontFamily": "WorkSans-Medium",
                  "fontSize": 14,
                  "marginLeft": 6,
                }
              }
            >
              Call
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "opacity": 0,
        }
      }
    >
      <View
        style={
          {
            "borderRadius": 16,
            "overflow": "hidden",
          }
        }
      >
        <View
          style={
            {
              "borderTopLeftRadius": 16,
              "borderTopRightRadius": 16,
              "overflow": "hidden",
            }
          }
        >
          <Image
            resizeMode="cover"
            source={
              {
                "uri": "https://example.com/image.jpg",
              }
            }
            style={
              {
                "height": 363,
                "width": 726,
              }
            }
          />
          <Text
            style={
              {
                "color": "#fff",
                "fontFamily": "WorkSans-SemiBold",
                "fontSize": 22,
                "fontWeight": "bold",
                "left": 0,
                "padding": 16,
                "position": "absolute",
                "textShadowColor": "#000",
                "textShadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "textShadowRadius": 6,
              }
            }
          >
            1
            .
          </Text>
          <Icon
            color="rgba(51,178,73,1)"
            name="favorite-border"
            size={24}
            style={
              {
                "padding": 16,
                "position": "absolute",
                "right": 0,
              }
            }
          />
        </View>
        <View
          style={
            {
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            }
          }
        >
          <View
            style={
              {
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                {
                  "color": "#000",
                  "flex": 1,
                  "fontFamily": "WorkSans-SemiBold",
                  "fontSize": 22,
                  "fontWeight": "bold",
                  "paddingBottom": 6,
                }
              }
            >
              Test Restaurant
            </Text>
            <Text
              style={
                {
                  "fontFamily": "WorkSans-SemiBold",
                  "fontSize": 22,
                }
              }
            >
              $$
            </Text>
          </View>
          <View
            style={
              {
                "flexDirection": "row",
              }
            }
          >
            <Text
              numberOfLines={1}
              style={
                {
                  "color": "rgba(128,128,128, 0.80)",
                  "flex": 1,
                  "fontFamily": "WorkSans-Regular",
                  "paddingRight": 4,
                }
              }
            >
              Italian, Pizza
              <View
                style={
                  {
                    "width": 4,
                  }
                }
              />
              <Icon
                color="rgba(51,178,73,1)"
                name="location-pin"
                size={12}
              />
              San Francisco
            </Text>
          </View>
          <View
            style={
              {
                "flexDirection": "row",
                "marginTop": 4,
              }
            }
          >
            <View
              style={
                {
                  "display": "flex",
                  "flexDirection": "row",
                }
              }
            >
              <MaterialIcons
                name="star"
                size={20}
                style={
                  {
                    "color": "#ffb300",
                  }
                }
              />
              <MaterialIcons
                name="star"
                size={20}
                style={
                  {
                    "color": "#ffb300",
                  }
                }
              />
              <MaterialIcons
                name="star"
                size={20}
                style={
                  {
                    "color": "#ffb300",
                  }
                }
              />
              <MaterialIcons
                name="star"
                size={20}
                style={
                  {
                    "color": "#ffb300",
                  }
                }
              />
              <MaterialIcons
                name="star-half"
                size={20}
                style={
                  {
                    "color": "#ffb300",
                  }
                }
              />
            </View>
            <Text
              style={
                {
                  "color": "rgba(128,128,128, 0.80)",
                  "fontFamily": "WorkSans-Regular",
                  "marginLeft": 8,
                }
              }
            >
              150
               Reviews
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
